osmhelper/osm_connector.py

```diff
@@ -0,0 +1,325 @@
+
+        stops = []
+
+        # Add ids for stops of this route variant
+        for stop_candidate in route_variant.members:
+            if stop_candidate.role == "platform":
+
+                if isinstance(stop_candidate, overpy.RelationNode):
+                    otype = "node"
+
+                elif isinstance(stop_candidate, overpy.RelationWay):
+                    otype = "way"
+
+                else:
+                    otype = "unknown"
+                    raise RuntimeError("Unknown type: " + str(stop_candidate))
+
+                stops.append(otype + "/" + str(stop_candidate.ref))
+
+        rv = Route(route_variant.id, None, None, stops, None, ref)
+        rv.add_shape(route_variant, query_result_set)
+        print rv
+        return rv
+
+    def _build_stop(self, stop, stop_type):
+        """Helper function to build a Stop object
+
+        Returns a initiated Stop object from raw data
+
+        """
+
+        # Make sure name is not empty
+        if 'name' not in stop.tags:
+            stop.tags['name'] = Stop.NO_NAME
+
+        # Ways don't have coordinates and they have to be calculated
+        if stop_type == "way":
+            (stop.lat, stop.lon) = Stop.get_center_of_nodes(stop.get_nodes())
+
+        s = Stop(stop.id, "node", stop.tags['name'], stop.lat, stop.lon)
+        print s
+        return s
+
+    def _query_routes(self):
+        """Helper function to query OpenStreetMap routes
+
+        Returns raw data on routes from OpenStreetMap
+
+        """
+        # Query relations of route variants, their masters and geometry
+        api = overpy.Overpass()
+        query_str = """(
+            /* Obtain route variants based on tags and bounding box */
+            relation%s(%s)->.routes;
+
+            /*  Query for related route masters */
+            relation[type=route_master](br.routes)->.masters;
+
+            /* Query for routes' geometry (ways and it's nodes) */
+            way(r.routes);
+            node(w);
+
+            /* Select all result sets  */
+            ( .routes;.masters;._; );
+
+            /* Return tags for elements and roles for relation members. */
+            );out body;""" % (self.tags, self.bbox)
+        return api.query(query_str)
+
+    def _query_stops(self):
+        """Helper function to query OpenStreetMap stops
+
+        Returns raw data on stops from OpenStreetMap
+
+        """
+        # Query stops with platform role from selected relations
+        api = overpy.Overpass()
+        query_str = """(
+            /* Obtain route variants based on tags and bounding box */
+            relation%s(%s);
+
+            /*  Query for related bus stop platforms */
+            node(r:"platform")->.nodes;
+            way(r:"platform");
+            node(w);
+
+            /* Select all result sets  */
+            ( .nodes;._; );
+
+            /* Return tags for elements */
+            );out body;""" % (self.tags, self.bbox)
+        return api.query(query_str)
```

Comment:
The query is actually looking for relations (the same relations used for the routes based on the tags) and then downloads all objects with the role "platform" of these relations (no matter which tags).

We could think about altering this to download all nodes which are tagged as possible stops. And then try to merge them with the route relations. But what about stops that are not properly tagged, what about stops that are from other networks, not subject of this. I think the common ground is, that we download osm route relations and it's stops and then verify them.